README.md
# ScalePageView 
![](https://img.shields.io/badge/license-GPL-orange.svg)

一个将相邻的Page进行缩放到同一个屏显示的控件,可以用来做以下用途:
1. 图片浏览
2. 单页的 Tab
3. 添加作为PageView标题栏
4. (**TODO**) 可以修改加入无限滚动和轮播功能,变成轮播图

### 效果图
<img src="http://qiniu.inrush.me/2018-08-14-imageDemo.png" width=286 /><img src="http://qiniu.inrush.me/2018-08-14-textImage.png" width=286 />

> 第一张图是将图片进行分页
> 第二张图是将文字进行分页 
 
### GIF 
> 可能会有点卡,GitHub显示不出来,可以到这里看[gif](http://qiniu.inrush.me/ScalePageView.gif)

![](http://qiniu.inrush.me/ScalePageView.gif)
### 使用
[Flutter Pub](https://pub.dartlang.org/packages/scale_page_view#-readme-tab-)

Add this to your package's pubspec.yaml file:
``` 
dependencies:
  scale_page_view: ^1.0.3
```
You can install packages from the command line:
```
$ flutter packages get
```
Now in your Dart code, you can use:

```
import 'package:scale_page_view/scale_page_view.dart';
```

``` dart
可用属性:
/// PageController
controller,
/// Page children
@required this.children,
/// type: ValueChanged<double> 
/// 页面滚动时触发,接收参数为 pageController.position.pixels
onPageChanging,
/// type: ValueChanged<int>
/// 页面变化时触发,接收的参数为改变后的新页面
onPageChange,
/// 当前页点击回调
onCurrentTabTap,
/// type: List<Widget>
/// 每一页的背景,length 必须等于 children.length
backgrounds,
/// type: Widget
/// 统一设置每一页的背景
background,
/// 设置是否显示指示器
indicator: true,
/// 设置指示器的颜色
indicatorColor: Colors.black,
/// 相邻的Page点击的时候是否要跳转
pageTapChange: true,
/// 当前页占屏幕宽度比例
pageRatio: 0.5,
/// 未选中页的缩放比,影响页面之间的间距
scaleRatio: 0.2,
/// 相邻页面的不透明度
opacityRatio: 0.5,
/// 页面的上下的Padding
paddingTB: 23.0

new Container(
    height: 200.0,
    child: new ScalePageView(
        key: _pageKey,
        children: images,
        backgrounds: backgrounds,
        onPageChanging: (value) {
            if (_pageController.position.pixels != value) {
                // _pageController 为另外一个PageView的Controller
                // 这样可以使ScalePageView移动的时候,另外一个PageView跟着动
                // 若要两个联动,可在另外一个PageView中添加NotificationListener,
                // 监听另外一个PageView滚动,然后通过GlobalKey<ScalePageViewState>,即上面设置的_pageKey,
                // 调用 _pageKey.currentState.jumpToWithoutSettling(
                // _pageController.position.pixels) 进行联动
                // 详细的实现请看 example 中的例子
                 _pageController.position.jumpToWithoutSettling(value);
            }
        },
        indicatorColor: Colors.white,
        pageRatio: 0.7,
        scaleRatio: 0.1,
        indicator: false,
        onPageChange: (index) {},
))
```
### 主要特性
1. 在一屏中可以看到相邻两屏的内容
2. 可以设置Widget作为背景
3. 可以设置一屏的占比,相邻两屏的透明度以及缩放比
4. 可以和其他PageView进行联动
5. 内置一个横线作为指示器

### TODO
1. 设置无限循环滚动
2. 加入轮播功能,可替代作为轮播图

### 参考
[Official Example - Flutter Gallery](https://github.com/flutter/flutter/tree/master/examples/flutter_gallery)

[Realank/flutter_card_navi](https://github.com/Realank/flutter_card_navi)



CHANGELOG.md
## [0.0.1] - The first release.

### Features
* Scale Multiple Page At One Page.

## [1.0.0] - Official release.

No Change

## [1.0.1] - (Minor edits) Add Property

add ScrollPhysics property

## [1.0.2] - (Minor edits) Fix bug

fix example recreate bug

## [1.0.3] - (Minor edits) Fix bug

* fix pageTapChange invalid bug
* delete pageTapOffset attributes


lib/scale_page_view.dart
library scale_page_view;

import 'package:flutter/material.dart';
import 'dart:math' as math;
const double kPageIndicatorWidth = 32.0;

class ScalePageView extends StatefulWidget {
  ScalePageView(
      {Key key,
      this.controller,
      @required this.children,
      this.onPageChanging,
      this.onPageChange,
      this.backgrounds,
      this.background,
      this.indicator: true,
      this.indicatorColor: Colors.black,
      this.pageTapChange: true,
      this.pageRatio: 0.5,
      this.scaleRatio: 0.2,
      this.opacityRatio: 0.5,
      this.paddingTB: 23.0,
      this.physics,
      this.onCurrentTabTap})
      : assert(children != null && children.length > 0),
        assert(pageRatio <= 1.0 && pageRatio > 0),
        assert(scaleRatio <= 1.0 && scaleRatio > 0),
        assert(opacityRatio <= 1.0 && opacityRatio > 0),
        assert(background == null || backgrounds == null),
        super(key: key);

  final List<Widget> children;
  final ValueChanged<double> onPageChanging;
  final ValueChanged<int> onPageChange;
  final List<Widget> backgrounds;
  final PageController controller;
  final Widget background;
  final bool indicator;
  final Color indicatorColor;
  final bool pageTapChange;

  final double pageRatio;
  final double scaleRatio;
  final double opacityRatio;
  final double paddingTB;
  final ScrollPhysics physics;
  final ValueChanged<int> onCurrentTabTap;

  @override
  State<StatefulWidget> createState() {
    return new ScalePageViewState();
  }
}

class ScalePageViewState extends State<ScalePageView> {
  PageController _pageController;
  ValueNotifier<double> selectedIndex = new ValueNotifier(0.0);
  List<Widget> backgrounds;

  /// 在使用Key调用该处的时候,切记不能在最外层使用StatelessWidget
  /// 这样可能会时Key发生变化,然后造成State重建
  /// The purpose of a stateful widget is that they don't recreate their state object every time
  /// they are built - otherwise they would be the same as a regular Widget.
  /// A state object is recreated if the widget's runtimeType or key is different.
  /// Otherwise the lifecycle didUpdateWidget is used to notify the State object that the configuration changed.
  @Deprecated('This will lead to bugs.')
  void jumpToWithoutSettling(double value) {
    if (_pageController.position.pixels != value) {
      _pageController.position.jumpToWithoutSettling(value);
    }
  }

  bool _handlePageNotification(ScrollNotification notification) {
    if (notification.depth == 0 && notification is ScrollUpdateNotification) {
      selectedIndex.value = _pageController.page;
      if (widget.onPageChanging != null) {
        widget.onPageChanging(_pageController.position.pixels);
      }
    }
    return false;
  }

  @override
  void initState() {
    super.initState();
    _pageController = widget.controller ?? new PageController();
    backgrounds = widget.backgrounds ?? [];
    if (backgrounds.length == 0) {
      Widget background = widget.background ??
          new Container(
            color: Colors.white,
          );
      for (int i = 0; i < widget.children.length; i++) {
        backgrounds.add(background);
      }
    }
  }

  Iterable<Widget> _allItems(List<Widget> pages) {
    final List<Widget> headings = <Widget>[];
    for (int index = 0; index < pages.length; index++) {
      headings.add(new Container(
        child: new ClipRect(
          child: new _AllPagesView(
            sectionIndex: index,
            selectedIndex: selectedIndex,
            pages: pages,
            backgrounds: backgrounds,
            pageTapChange: widget.pageTapChange,
            indicator: widget.indicator,
            indicatorColor: widget.indicatorColor,
            pageRatio: widget.pageRatio,
            scaleRatio: widget.scaleRatio,
            opacityRatio: widget.opacityRatio,
            paddingTB: widget.paddingTB,
            onTabTap: (index) {
              if (index == _pageController.page.ceil() && widget.onCurrentTabTap != null) {
                widget.onCurrentTabTap(index);
              } else if (widget.pageTapChange) {
                _pageController.animateToPage(index,
                    duration: new Duration(milliseconds: 500),
                    curve: Curves.fastOutSlowIn);
              }
            },
          ),
        ),
      ));
    }
    return headings;
  }

  @override
  Widget build(BuildContext context) {
    return new NotificationListener(
        onNotification: _handlePageNotification,
        child: new PageView(
          controller: _pageController,
          children: _allItems(widget.children),
          onPageChanged: widget.onPageChange,
          physics: widget.physics,
        ));
  }
}

class _AllPagesView extends AnimatedWidget {
  _AllPagesView(
      {Key key,
      this.sectionIndex,
      @required this.selectedIndex,
      @required this.pages,
      this.backgrounds,
      this.onTabTap,
      this.pageTapChange,
      this.pageTapOffset,
      this.indicator,
      this.indicatorColor,
      this.pageRatio,
      this.scaleRatio,
      this.opacityRatio,
      this.paddingTB})
      : assert(selectedIndex != null),
        assert(selectedIndex.value >= 0.0 &&
            selectedIndex.value < pages.length.toDouble()),
        assert(backgrounds == null || backgrounds.length == pages.length),
        super(key: key, listenable: selectedIndex);

  final int sectionIndex;
  final ValueNotifier<double> selectedIndex;
  final List<Widget> pages;
  final List<Widget> backgrounds;
  final ValueChanged<int> onTabTap;
  final bool pageTapChange;
  final double pageTapOffset;
  final bool indicator;
  final Color indicatorColor;
  final double pageRatio;
  final double scaleRatio;
  final double opacityRatio;
  final double paddingTB;

  double _selectedIndexDelta(int index) {
    return (index.toDouble() - selectedIndex.value).abs().clamp(0.0, 1.0);
  }

  /// calculate page and indicator opacity or scale
  /// when the index equal current page, will return 1.0
  double _calculateScale(int index) {
    return 1.0 - _selectedIndexDelta(index) * scaleRatio;
  }

  double _calculateOpacity(int index) {
    return 1.0 - _selectedIndexDelta(index) * opacityRatio;
  }

  void changePage(TapUpDetails details, double width, int index) {
    if (pageTapChange && onTabTap != null) {
      final double xOffset = details.globalPosition.dx;
      final double centerX = width / 2.0;
      int newPageIndex = index;
      if (xOffset - centerX > pageTapOffset &&
          newPageIndex < pages.length - 1) {
        newPageIndex++;
      } else if (centerX - xOffset > pageTapOffset && newPageIndex > 0) {
        newPageIndex--;
      }
      onTabTap(newPageIndex);
    }
  }

  @override
  Widget build(BuildContext context) {
    return new LayoutBuilder(builder: (context, size) {
      final List<Widget> children = [];
      // put all the backgrounds at the bottom
      for (int index = 0; index < pages.length; index++) {
        children.add(new LayoutId(
          id: 'bg$index',
          child: new GestureDetector(
              behavior: HitTestBehavior.opaque,
//              onTapUp: (TapUpDetails details) {
//                changePage(details, size.maxWidth, index);
//              },
              child: new Container(
                child: backgrounds == null ? null : backgrounds[index],
              )),
        ));
      }
      for (int index = 0; index < pages.length; index++) {
        children.add(new LayoutId(
          id: 'page$index',
          child: new GestureDetector(
            onTapUp: (TapUpDetails details) {
              if (onTabTap != null) {
                onTabTap(index);
              }
            },
            child: new _PageWrapper(
                child: pages[index],
                scale: _calculateScale(index),
                opacity: _calculateOpacity(index)),
          ),
        ));
        if (indicator) {
          children.add(new LayoutId(
            id: 'indicator$index',
            child: new _PageIndicator(
              opacity: _calculateOpacity(index),
              color: indicatorColor,
            ),
          ));
        }
      }
      return new CustomMultiChildLayout(
        delegate: new _AllPagesLayout(
            translation: new Alignment(
                (selectedIndex.value - sectionIndex) * 2.0 - 1.0, -1.0),
            selectedIndex: selectedIndex.value,
            pageRatio: pageRatio,
            pageCount: pages.length,
            paddingTB: paddingTB),
        children: children,
      );
    });
  }
}

class _AllPagesLayout extends MultiChildLayoutDelegate {
  _AllPagesLayout(
      {this.translation,
      this.selectedIndex,
      this.pageCount,
      this.pageRatio,
      double paddingTB})
      : padding = paddingTB * 2;

  final Alignment translation;
  final double selectedIndex;
  final int pageCount;
  final double pageRatio;
  final double padding;

  @override
  void performLayout(Size size) {
    assert(size.height > padding);
    final double itemWidth = size.width;
    final double itemHeight = size.height;
    // 计算在当前位置下的offset,其主要作用是利用Alignment,使得item进行相应的偏移
    // 因为titleWidth和itemWidth不一致,需要用这个达到减缓位移速度的效果,最终使得title还是在中间
    // 用来抵消掉由于pageView移动产生的影响
    final Offset offset = translation.alongSize(size);
    double itemX = -(selectedIndex * itemWidth);

    final double pageAreaWidth = pageRatio * itemWidth;
    double itemPageAreaX =
        (size.width - pageAreaWidth) / 2.0 - selectedIndex * pageAreaWidth;

    final double indicatorAreaWidth = pageAreaWidth;

    // 最后的减法是用来产生移动的
    double itemIndicatorAreaX = (size.width - indicatorAreaWidth) / 2.0 -
        selectedIndex * indicatorAreaWidth;

    for (int index = 0; index < pageCount; index++) {
      final Rect itemRect =
          new Rect.fromLTWH(itemX, 0.0, itemWidth, itemHeight);
      final Rect pageAreaRect = new Rect.fromLTWH(
          itemPageAreaX, 0.0, pageAreaWidth, itemHeight - padding);
      // layout background
      layoutChild('bg$index', new BoxConstraints.tight(itemRect.size));
      positionChild('bg$index', itemRect.topLeft + offset);

      // layout title
      final Size pageSize = layoutChild(
          'page$index', new BoxConstraints.loose(pageAreaRect.size));
      final double pageY = itemRect.centerLeft.dy - pageSize.height / 2.0;
      final double pageX =
          itemPageAreaX + (pageAreaWidth - pageSize.width) / 2.0;
      final Offset pageOrigin = new Offset(pageX, pageY);
      positionChild('page$index', pageOrigin + offset);

      // layout indicator
      if (hasChild('indicator$index')) {
        final Size indicatorSize = layoutChild(
            'indicator$index', new BoxConstraints.loose(pageAreaRect.size));
        final Rect pageRect =
            new Rect.fromPoints(pageOrigin, pageSize.bottomRight(pageOrigin));

        // 使得indicator在title正下方
        final double indicatorY = pageRect.bottomCenter.dy + 14.0;
        final double indicatorX = itemIndicatorAreaX +
            (indicatorAreaWidth - indicatorSize.width) / 2.0;

        final Offset indicatorOrigin = new Offset(indicatorX, indicatorY);
        positionChild('indicator$index', indicatorOrigin + offset);
      }

      itemX += itemWidth;
      itemPageAreaX += pageAreaWidth;
      itemIndicatorAreaX += indicatorAreaWidth;
    }
  }

  @override
  bool shouldRelayout(_AllPagesLayout oldDelegate) {
    return pageCount != oldDelegate.pageCount ||
        selectedIndex != oldDelegate.selectedIndex;
  }
}

class _PageWrapper extends StatelessWidget {
  const _PageWrapper({
    Key key,
    @required this.child,
    @required this.scale,
    @required this.opacity,
  })  : assert(child != null),
        assert(scale != null),
        assert(opacity != null && opacity >= 0.0 && opacity <= 1.0),
        super(key: key);

  final Widget child;
  final double scale;
  final double opacity;

  @override
  Widget build(BuildContext context) {
    return new Container(
      child: new Opacity(
        opacity: opacity,
        child: new Transform(
          transform: new Matrix4.identity()..scale(scale),
          alignment: Alignment.center,
          child: child,
        ),
      ),
    );
  }
}

class _PageIndicator extends StatelessWidget {
  const _PageIndicator({Key key, this.opacity: 1.0, this.color: Colors.white})
      : super(key: key);

  final double opacity;
  final Color color;

  @override
  Widget build(BuildContext context) {
    return new IgnorePointer(
      child: new Container(
        width: kPageIndicatorWidth,
        height: 3.0,
        color: color.withOpacity(opacity),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:scale_page_view/scale_page_view.dart';

void main() {
  MaterialPageRoute.debugEnableFadingRoutes = true;
  runApp(new MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Widget _buildItem(String name, WidgetBuilder builder) {
    return new ListTile(
      title: new GestureDetector(
        onTap: () {
          Navigator.of(context).push(new MaterialPageRoute(builder: builder));
        },
        child: new Text(name),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text("ScalePageView"),
        elevation: 0.5,
      ),
      body: new ListView(
        children: <Widget>[
          new Divider(),
          _buildItem("Example1 Image Page View", (context) {
            return new Example1();
          }),
          new Divider(),
          _buildItem("Example2 Text Page View", (context) {
            return new Example2();
          }),
          new Divider(),
        ],
      ),
    );
  }
}

const double _kBar1Height = 300.0;
const double _kBar2Height = 100.0;

List<Widget> images = [
  new Image.network(
    "https://petbird.tw/wp-content/uploads/2013/06/pet105.jpg",
  ),
  new Image.network(
    "https://petbird.tw/wp-content/uploads/2013/06/pet105.jpg",
  ),
  new Image.network(
    "https://petbird.tw/wp-content/uploads/2013/06/pet105.jpg",
  ),
  new Image.network(
    "https://petbird.tw/wp-content/uploads/2013/06/pet105.jpg",
  ),
  new Image.network(
    "https://petbird.tw/wp-content/uploads/2013/06/pet105.jpg",
  ),
];

Widget _buildBackground(Color leftColor, Color rightColor) {
  return new Container(
    decoration: new BoxDecoration(
      gradient: new LinearGradient(
        begin: Alignment.centerLeft,
        end: Alignment.centerRight,
        colors: <Color>[
          leftColor,
          rightColor,
        ],
      ),
    ),
  );
}

List<Widget> backgrounds = [
  _buildBackground(const Color(0xFF3B5F8F), const Color(0xFF8266D4)),
  _buildBackground(const Color(0xFF8266D4), const Color(0xFFF95B57)),
  _buildBackground(const Color(0xFFF95B57), const Color(0xFFF3A646)),
  _buildBackground(const Color(0xFFF3A646), const Color(0xFF3B5F8F)),
  _buildBackground(const Color(0xFF3B5F8F), const Color(0xFF8266D4)),
];
const TextStyle sectionTitleStyle = const TextStyle(
  fontFamily: 'Raleway',
  inherit: false,
  fontSize: 20.0,
  fontWeight: FontWeight.w500,
  color: Colors.white,
  textBaseline: TextBaseline.alphabetic,
);
List<String> texts = [
  "西伯利亚雪橇犬",
  "西伯利亚雪橇犬",
  "西伯利亚雪橇犬",
  "西伯利亚雪橇犬",
  "西伯利亚雪橇犬",
];
List<String> descs = [
  "       西伯利亚雪橇犬（俄语：Сибирский хаски，英语：Siberian husky），常见别名哈士奇，昵称为二哈。西伯利亚雪橇犬体重介于雄犬20-27公斤，雌犬16-23公斤，身高大约雄犬肩高53-58厘米，雌犬51-56厘米，是一种中型犬。\n       西伯利亚雪橇犬是原始的古老犬种，在西伯利亚东北部、格陵兰南部生活。哈士奇名字的由来，是源自其独特的嘶哑声。哈士奇性格多变，有的极端胆小，也有的极端暴力，进入大陆和家庭的哈士奇，都已经没有了这种极端的性格，比较温顺，是一种流行于全球的宠物犬。与金毛犬、拉布拉多并列为三大无攻击型犬类。被世界各地广泛饲养，并在全球范围内，有大量该犬种的赛事。",
  "       西伯利亚雪橇犬（俄语：Сибирский хаски，英语：Siberian husky），常见别名哈士奇，昵称为二哈。西伯利亚雪橇犬体重介于雄犬20-27公斤，雌犬16-23公斤，身高大约雄犬肩高53-58厘米，雌犬51-56厘米，是一种中型犬。\n       西伯利亚雪橇犬是原始的古老犬种，在西伯利亚东北部、格陵兰南部生活。哈士奇名字的由来，是源自其独特的嘶哑声。哈士奇性格多变，有的极端胆小，也有的极端暴力，进入大陆和家庭的哈士奇，都已经没有了这种极端的性格，比较温顺，是一种流行于全球的宠物犬。与金毛犬、拉布拉多并列为三大无攻击型犬类。被世界各地广泛饲养，并在全球范围内，有大量该犬种的赛事。",
  "       西伯利亚雪橇犬（俄语：Сибирский хаски，英语：Siberian husky），常见别名哈士奇，昵称为二哈。西伯利亚雪橇犬体重介于雄犬20-27公斤，雌犬16-23公斤，身高大约雄犬肩高53-58厘米，雌犬51-56厘米，是一种中型犬。\n       西伯利亚雪橇犬是原始的古老犬种，在西伯利亚东北部、格陵兰南部生活。哈士奇名字的由来，是源自其独特的嘶哑声。哈士奇性格多变，有的极端胆小，也有的极端暴力，进入大陆和家庭的哈士奇，都已经没有了这种极端的性格，比较温顺，是一种流行于全球的宠物犬。与金毛犬、拉布拉多并列为三大无攻击型犬类。被世界各地广泛饲养，并在全球范围内，有大量该犬种的赛事。",
  "       西伯利亚雪橇犬（俄语：Сибирский хаски，英语：Siberian husky），常见别名哈士奇，昵称为二哈。西伯利亚雪橇犬体重介于雄犬20-27公斤，雌犬16-23公斤，身高大约雄犬肩高53-58厘米，雌犬51-56厘米，是一种中型犬。\n       西伯利亚雪橇犬是原始的古老犬种，在西伯利亚东北部、格陵兰南部生活。哈士奇名字的由来，是源自其独特的嘶哑声。哈士奇性格多变，有的极端胆小，也有的极端暴力，进入大陆和家庭的哈士奇，都已经没有了这种极端的性格，比较温顺，是一种流行于全球的宠物犬。与金毛犬、拉布拉多并列为三大无攻击型犬类。被世界各地广泛饲养，并在全球范围内，有大量该犬种的赛事。",
  "       西伯利亚雪橇犬（俄语：Сибирский хаски，英语：Siberian husky），常见别名哈士奇，昵称为二哈。西伯利亚雪橇犬体重介于雄犬20-27公斤，雌犬16-23公斤，身高大约雄犬肩高53-58厘米，雌犬51-56厘米，是一种中型犬。\n       西伯利亚雪橇犬是原始的古老犬种，在西伯利亚东北部、格陵兰南部生活。哈士奇名字的由来，是源自其独特的嘶哑声。哈士奇性格多变，有的极端胆小，也有的极端暴力，进入大陆和家庭的哈士奇，都已经没有了这种极端的性格，比较温顺，是一种流行于全球的宠物犬。与金毛犬、拉布拉多并列为三大无攻击型犬类。被世界各地广泛饲养，并在全球范围内，有大量该犬种的赛事。",
];

class Example1 extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return new Example1State();
  }
}

/// Use State,avoid reinitializing variables because of refresh
class Example1State extends State<Example1> {
  final PageController _pageController = new PageController();
  final PageController _tabController = new PageController();

  @override
  Widget build(BuildContext context) {
    return new Material(
      child: new Column(
        children: <Widget>[
          new Container(
              height: _kBar1Height,
              child: new ScalePageView(
                controller: _tabController,
                children: images,
                pageTapChange: true,
//              backgrounds: backgrounds,
                onPageChanging: (value) {
                  if (_pageController.position.pixels != value) {
                    _pageController.position.jumpToWithoutSettling(value);
                  }
                },
                onCurrentTabTap: (index){
                  debugPrint("touch $index");
                },
                indicatorColor: Colors.white,
                pageRatio: 0.7,
                scaleRatio: 0.1,
                indicator: false,
                onPageChange: (index) {},
              )),
          new Expanded(
              child: new NotificationListener(
                  onNotification: (notification) {
                    if (notification is ScrollUpdateNotification) {
                      if (_tabController.position.pixels !=
                          _pageController.position.pixels) {
                        _tabController.position.jumpToWithoutSettling(
                            _pageController.position.pixels);
                      }
                    }
                  },
                  child: new PageView.builder(
                    controller: _pageController,
                    itemBuilder: (context, index) {
                      return new Container(
                          color: Colors.white,
                          padding: const EdgeInsets.all(20.0),
                          child: new SingleChildScrollView(
                            child: new Column(
                              children: <Widget>[
                                new Row(
                                  mainAxisAlignment: MainAxisAlignment.center,
                                  children: <Widget>[
                                    new Text(
                                      texts[index],
                                      style: new TextStyle(
                                          fontSize: 23.0,
                                          color: Color(0xff333333)),
                                    )
                                  ],
                                ),
                                new Padding(
                                  padding: const EdgeInsets.only(top: 20.0),
                                  child: new RichText(
                                    text: new TextSpan(
                                        text: descs[index],
                                        style: new TextStyle(
                                            fontSize: 18.0,
                                            color: Color(0xff999999))),
                                  ),
                                )
                              ],
                            ),
                          ));
                    },
                    itemCount: images.length,
                  )))
        ],
      ),
    );
  }
}

class Example2 extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return new Example2State();
  }
}

class Example2State extends State<Example2> {
  final PageController _pageController = new PageController();
  final PageController _tabController = new PageController();

  @override
  Widget build(BuildContext context) {
    return new Material(
      child: new Column(
        children: <Widget>[
          new Container(
              height: _kBar2Height,
              child: new ScalePageView(
                controller: _tabController,
                children: texts.map((String text) {
                  return new Text(
                    text,
                    style: sectionTitleStyle,
                  );
                }).toList(),
                backgrounds: backgrounds,
                onPageChanging: (value) {
                  if (_pageController.position.pixels != value) {
                    _pageController.position.jumpToWithoutSettling(value);
                  }
                },
                indicatorColor: Colors.white,
                pageRatio: 0.5,
                scaleRatio: 0.3,
                indicator: true,
                onPageChange: (index) {},
              )),
          new Expanded(
              child: new NotificationListener(
                  onNotification: (notification) {
                    if (notification is ScrollUpdateNotification) {
                      if (_tabController.position.pixels !=
                          _pageController.position.pixels) {
                        _tabController.position.jumpToWithoutSettling(
                            _pageController.position.pixels);
                      }
                    }
                  },
                  child: new PageView.builder(
                    controller: _pageController,
                    itemBuilder: (context, index) {
                      return new Container(
                        color: Colors.white,
                        padding: const EdgeInsets.all(20.0),
                        child: new SingleChildScrollView(
                          child: new Column(
                            children: <Widget>[
                              new Padding(
                                padding: const EdgeInsets.only(top: 20.0),
                                child: images[index],
                              ),
                              new Padding(
                                padding: const EdgeInsets.only(top: 20.0),
                                child: new RichText(
                                  text: new TextSpan(
                                      text: descs[index],
                                      style: new TextStyle(
                                          fontSize: 18.0,
                                          color: Color(0xff999999))),
                                ),
                              )
                            ],
                          ),
                        ),
                      );
                    },
                    itemCount: images.length,
                  )))
        ],
      ),
    );
  }
}
